<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

_Author: Matt Brems_

### Required Preparation

In order to run SQL queries within Python/this Jupyter notebook, you will need to install the following:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you get an error trying to install `psycopg2`, you can install a stand-alone package (sufficient for everything you'll need to complete this lab) by entering the following anywhere in the terminal: 

```bash 
pip install psycopg2-binary
```

(If `pip` doesn't work for you, try `pip3 install psycopg2-binary`.)

### Instructions

In this lab, you're going to have ten prompts. Each prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to get you to practice your SQL - an **important** skill in data science! While it is possible to get these answers using Pandas, it should be your goal to do everything only in SQL.

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The schema for this database is here:

<img src="./images/schema.png" alt="schema" width="750"/>

__Query Order:__
1. SELECT
1. FROM
1. WHERE
1. GROUP BY 
1. HAVING
1. ORDER BY
1. LIMIT

In [1]:
# Import Pandas and Create_Engine

import pandas as pd
from sqlalchemy import create_engine

# Set up the engine to access the data.
engine = create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [3]:
sql = '''
SELECT table_name FROM information_schema.tables
WHERE table_schema = 'public'
'''
pd.read_sql_query(sql, engine)

table_name
0   counties
1   products
2      sales
3     stores

In [2]:
# Use a SQL query to find the answer to the above prompt.

sql = 'SELECT Count(*) FROM stores'

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**: 1973

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = '''SELECT * 
FROM stores 
ORDER BY name asc 
LIMIT 4'''

pd.read_sql_query(sql, engine)

store                                name store_status  \
0   4986  218 Fuel Express & Chubby's Liquor            A   
1   3047          3047 Cub Foods / Iowa City            I   
2   3060               3060 Cub Foods / Ames            I   
3   3061         3061 Cub Foods / Sioux City            I   

                                       store_address address_info  
0  68 Monroe St\nFloyd, IA 50435\n(43.12249790500...         None  
1  855 Highway 1 West\nIowa City, IA 522460000\n(...         None  
2  3121 Grand Ave\nAmes, IA 500100000\n(42.053458...         None  
3  1732 Hamilton Blvd\nSioux City, IA 511030000\n...         None

In [4]:
# Use a SQL query to find the answer to the above prompt.
# windows function

sql = '''
Select B.name from
(SELECT *, row_number() over (order by name) as num FROM stores) B 
where B.num=4; 
'''

pd.read_sql_query(sql, engine)

name
0  3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT vendor, category_name, btl_price, bottle_qty 
FROM sales 
WHERE category_name = 'IMPORTED VODKA'
ORDER BY vendor asc 
LIMIT 3
"""

pd.read_sql_query(sql, engine)

vendor   category_name btl_price  bottle_qty
0  Adamba Imports Int'l Inc.  IMPORTED VODKA     $8.99          12
1      A Hardy / U.S.A. Ltd.  IMPORTED VODKA    $28.50           3
2      A Hardy / U.S.A. Ltd.  IMPORTED VODKA    $28.50           6

__Prompt 3 Answer:__

|Rank|Bottle price|Bottle quantity|
|:-:|:-:|:-:|
|1|8.99|12|
|2|28.50|6|
|3|28.50|6|

### Prompt 4

What is the total amount of liquor sold in February 2015?

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT sum(total) 
FROM sales
WHERE date > '2015-01-31' and date < '2015-03-01'
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**: \$21,295,350.55

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. Rename the columns to make sure there isn't any confusion as to what they mean!

In [6]:
# Use a SQL query to find the answer to the above prompt.

sql = """SELECT vendor, COUNT(vendor) as number_of_transactions, SUM(bottle_qty) as bottles_sold, SUM(total) as revenue 
FROM sales
WHERE category_name = 'IMPORTED VODKA'
GROUP BY vendor
ORDER BY count(vendor) desc
LIMIT 1
"""

pd.read_sql_query(sql, engine)

number_of_transactions                            vendor  bottles_sold  \
0                   42338  Pernod Ricard USA/Austin Nichols        403450   

      revenue  
0  7880365.88

**Prompt 5 Answer**: 

|number_of_transactions|vendor|bottles_sold|revenue|
|---|:-:|---|---|
|42,338|Pernod Ricard USA/Austin Nichols|403,450|7,880,365.88|

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of extra money charged for a product over the cost to make that product. (You can think of the markup as the profit for that product.)

In Iowa, the law states the minimum price at which liquor may be sold.

Calculate the markup (name this `Markup`) that shows the difference between the `btl_price` and the `state_btl_cost`. 

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT *, btl_price - state_btl_cost as 'markup' 
FROM sales
LIMIT 2
"""

pd.read_sql_query(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   
1 2014-11-13              None   2635            82  Scott  1032200   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   
1  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34076   

       description  pack  liter_size state_btl_cost btl_price  bottle_qty  \
0    Absolut Pears    12         750         $11.49    $17.24           4   
1  Absolut Vanilia    12         750         $11.49    $17.24          12   

    total markup  
0   68.96  $5.75  
1  206.88  $5.75

### Prompt 7

Calculate the "Percentage Markup" by dividing markup by the state bottle cost. (Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum.)

After calculating the "Percentage Markup," calculate the average percentage markup per **store**. Which store has the largest average markup, and what is that average markup?

In [8]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT store, AVG((btl_price - state_btl_cost)/state_btl_cost) as percentage_markup  
FROM sales
GROUP BY convenience_store, store
ORDER BY percentage_markup desc
LIMIT 1
"""

pd.read_sql_query(sql, engine)

store  percentage_markup
0   4013           0.525471

**Prompt 7 Answer**: 

|store|percentage_markup|
|---|---|
|4013|52.5471%|

### Prompt 8

Building off of your last query, retrieve the five stores with the highest average markup percentage. Along with the store IDs and average markup percentage, please return the minimum and maximum markup percentage for those stores, as well as the names and addresses of those store.

What are the names of the five stores?

In [9]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT ss.store, st.name, st.store_address, 
AVG((ss.btl_price - ss.state_btl_cost)/ss.state_btl_cost) as 'percentage_markup', 
MIN((ss.btl_price - ss.state_btl_cost)/ss.state_btl_cost), 
MAX((ss.btl_price - ss.state_btl_cost)/ss.state_btl_cost)  
FROM sales ss, stores st
WHERE ss.store = st.store
GROUP BY ss.store, st.store
ORDER BY 'percentage_markup' desc
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store                        name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   
4   3660   Wal-Mart 2935 / Knoxville   

                                       store_address  percentage_markup  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...           0.525471   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...           0.521499   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...           0.518903   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....           0.515094   
4  814 W Bell Ave\nKnoxville, IA 501380000\n(41.3...           0.514591   

        min       max  
0  0.498571  6.708763  
1  0.498462  6.708763  
2  0.498462  6.708763  
3  0.498462  6.708763  
4  0.498462  6.708763

**Prompt 8 Answer**:

|store id|store names|
|---|---|
|4013|Wal-Mart 0841 / Tipton|
|4024|Wal-Mart 1546 / Iowa Falls|
|4266|Wal-Mart 1683 / Shenandoah|
|3833|Wal-Mart 3394 / Atlantic|
|3660|Wal-Mart 2935 / Knoxville|

### Prompt 9

Which nine counties sell the highest number of bottles of liquor per capita (per person)? Calculate the average bottle size of the bottles of liquor sold by these counties rounded to the nearest first decimal place. Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

**_BONUS FUN_**: This question can (also) be solved with a SQL subquery - that is, a query that queries a query! If you'd like to give this a try, you can read more about it [here](https://www.dofactory.com/sql/subquery). Otherwise, keep an eye out for the solution key!

In [10]:
# Subquery

sql = """
SELECT s.county, 
SUM(s.bottle_qty)/c.population as btl_per_capita, 
ROUND(AVG((p.bottle_size*s.bottle_qty)/s.bottle_qty),1) as avg_btl_size
FROM sales s, counties c, products p
WHERE s.county = c.county AND s.item = p.item_no
GROUP BY s.county, c.county
ORDER BY btl_per_capita desc
LIMIT 9
"""

pd.read_sql_query(sql, engine)

county  btl_per_capita  avg_btl_size
0      Dickinson              22         974.1
1           Polk              15         892.6
2     Black Hawk              15         862.9
3    Cerro Gordo              14         986.4
4          Scott              13         871.6
5        Johnson              13         895.4
6           Linn              12         886.3
7  Pottawattamie              11         893.7
8        Kossuth              11        1041.0

In [11]:
# Performing a query on a result from another query
# referenced 
# https://stackoverflow.com/questions/949465/performing-a-query-on-a-result-from-another-query

sql = """SELECT county, MAX(avg_btl_size) FROM
    (
    SELECT s.county, 
    SUM(s.bottle_qty)/c.population as btl_per_capita, 
    ROUND(AVG((p.bottle_size*s.bottle_qty)/s.bottle_qty),1) as avg_btl_size
    FROM sales s, counties c, products p
    WHERE s.county = c.county AND s.item = p.item_no
    GROUP BY s.county, c.county
    ORDER BY btl_per_capita desc
    LIMIT 9
) AS SUBQUERY
GROUP BY county
ORDER BY MAX(avg_btl_size) desc
LIMIT 1
"""

pd.read_sql_query(sql, engine)

county     max
0  Kossuth  1041.0

**Prompt 9 Answer**: Of the nine counties selling the highest number of bottles of liquor per capita, Kossuth county sells (on average) the largest bottle size 1041.

### Prompt 10 (advanced but required!)

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions have a value of `total` that is not equal to `btl_price` time `bottle_qty`?

> There will be a type error as well! We cannot compare type `money` to type `real`. We have to convert `total` to type `money` so that we can compare the two directly. [This link](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#syntax) may be helpful to you!

In [12]:
# Checking if query works without condition

sql = """
SELECT bottle_qty, btl_price, 
bottle_qty*btl_price AS cal_t, 
CAST(total AS money) 
FROM sales
LIMIT 5
"""
pd.read_sql_query(sql, engine)

bottle_qty btl_price    cal_t    total
0           4    $17.24   $68.96   $68.96
1          12    $17.24  $206.88  $206.88
2          12    $17.24  $206.88  $206.88
3           6    $25.73  $154.38  $154.38
4           6     $7.50   $45.00   $45.00

In [13]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT bottle_qty, btl_price, 
bottle_qty*btl_price AS cal_t, 
CAST(total AS money) 
FROM sales
WHERE bottle_qty*btl_price <> CAST(total AS money)
LIMIT 5
"""
pd.read_sql_query(sql, engine)

Empty DataFrame
Columns: [bottle_qty, btl_price, cal_t, total]
Index: []

In [5]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT count(bottle_qty, btl_price, 
bottle_qty*btl_price AS cal_t, 
CAST(total AS money)) 
FROM sales
WHERE bottle_qty*btl_price <> CAST(total AS money)
LIMIT 5
"""
pd.read_sql_query(sql, engine)

OperationalError: (psycopg2.OperationalError) server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.

[SQL: 
SELECT count(bottle_qty, btl_price, 
bottle_qty*btl_price AS cal_t, 
CAST(total AS money)) 
FROM sales
WHERE bottle_qty*btl_price <> CAST(total AS money)
LIMIT 5
]
(Background on this error at: https://sqlalche.me/e/14/e3q8)

**Prompt 10 Answer**: None